### ISSUE TITLE
Python: Bug: Process Framework typing type hint error when data passed in event contains datetime

### ISSUE THREAD

**Issue by Druid-of-Luhn** [microsoft/semantic-kernel#9606]
**Describe the bug**

When an object containing a datetime field is passed as the data in an event between steps in the process framework, the code crashes with a type hint error from typing saying that datetime is not defined/imported.

**To Reproduce**

Run the following code snippet:

```python
"""Minimal example of Process Framework error with datetime."""

from __future__ import annotations

import asyncio
import datetime
import logging

from pydantic import BaseModel
from semantic_kernel import Kernel
from semantic_kernel.functions import kernel_function
from semantic_kernel.processes.kernel_process.kernel_process_event import KernelProcessEvent
from semantic_kernel.processes.kernel_process.kernel_process_step import KernelProcessStep
from semantic_kernel.processes.local_runtime import local_kernel_process
from semantic_kernel.processes.process_builder import ProcessBuilder


class Data(BaseModel):
    """Data passed between steps."""

    timestamp: datetime.datetime


class Step(KernelProcessStep):
    """Step that receives data with a datetime field."""

    @kernel_function
    async def process_event(self, data: Data) -> None:
        """Process an event containing data with a datetime field."""
        logging.info("Handled event with data %s", data)


async def run_process() -> None:
    """Run a Semantic Kernel Process Framework process with data containing a datetime."""
    process_builder = ProcessBuilder(name="Process")  # pyright: ignore [reportCallIssue]
    step = process_builder.add_step(Step)
    process_builder.on_input_event(event_id="event").send_event_to(
        target=step,
        parameter_name="data",
    )
    process = process_builder.build()

    data = Data(timestamp=datetime.datetime.now())  # noqa: DTZ005

    kernel = Kernel()
    await local_kernel_process.start(
        process=process,
        kernel=kernel,
        initial_event=KernelProcessEvent(id="event", data=data),
    )


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    asyncio.run(run_process())
```

The following error logs are produced:

```
Traceback (most recent call last):
  File "<project_path>\process_datetime_error.py", line 55, in <module>
    asyncio.run(run_process())
  File "<user_dir>\AppData\Local\miniconda3\Lib\asyncio\runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "<user_dir>\AppData\Local\miniconda3\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<user_dir>\AppData\Local\miniconda3\Lib\asyncio\base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "<project_path>\process_datetime_error.py", line 36, in run_process
    step = process_builder.add_step(Step)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\processes\process_builder.py", line 49, in add_step
    process_step_builder = ProcessStepBuilder(type=step_type, name=name, initial_state=initial_state, **kwargs)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\processes\process_step_builder.py", line 55, in __init__
    functions_dict = self.get_function_metadata_map(type, name, kwargs.get("kernel", None))
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\processes\process_step_builder.py", line 207, in get_function_metadata_map
    plugin_functions = kernel.add_plugin(step_instance, plugin_name)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\functions\kernel_function_extension.py", line 92, in add_plugin
    self.plugins[plugin_name] = KernelPlugin.from_object(
                                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\functions\kernel_plugin.py", line 263, in from_object
    KernelFunctionFromMethod(method=candidate, plugin_name=plugin_name)
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\functions\kernel_function_from_method.py", line 57, in __init__
    parameters = [KernelParameterMetadata(**param) for param in method.__kernel_function_parameters__]  # type: ignore
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\pydantic\main.py", line 212, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\functions\kernel_parameter_metadata.py", line 33, in form_schema
    inferred_schema = cls.infer_schema(type_object, type_, default_value, description)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\functions\kernel_parameter_metadata.py", line 50, in infer_schema
    schema = KernelJsonSchemaBuilder.build(type_object, description, structured_output)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\schema\kernel_json_schema_builder.py", line 57, in build
    return cls.build_model_schema(parameter_type, description, structured_output)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<project_path>\venv\Lib\site-packages\semantic_kernel\schema\kernel_json_schema_builder.py", line 83, in build_model_schema
    hints = get_type_hints(model, globals(), locals())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<user_dir>\AppData\Local\miniconda3\Lib\typing.py", line 2272, in get_type_hints
    value = _eval_type(value, base_globals, base_locals, base.__type_params__)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<user_dir>\AppData\Local\miniconda3\Lib\typing.py", line 415, in _eval_type
    return t._evaluate(globalns, localns, type_params, recursive_guard=recursive_guard)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<user_dir>\AppData\Local\miniconda3\Lib\typing.py", line 947, in _evaluate
    eval(self.__forward_code__, globalns, localns),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 1, in <module>
NameError: name 'datetime' is not defined. Did you forget to import 'datetime'?
```

**Expected behavior**

I would expect there to be no issue with passing an object containing a datetime field, or for there to be mention of the fact that it is not possible somewhere in the documentation or in an error message.

**Screenshots**
n/a

**Platform**
 - OS: Windows
 - IDE: gVim + PowerShell
 - Language: Python 3.12
 - Source: pip package version 1.13.0

**Additional context**
n/a

**Comment by moonbox3** [comment]
Thanks, @Druid-of-Luhn, looking now.

**Comment by moonbox3** [comment]
Again, thanks for filing this, @Druid-of-Luhn! I just investigated this and it was a deeper issue in SK, and not completely related to processes, although that's where we caught it, which is great.

The the root cause of the issue is that the `get_type_hints` function in the `kernel_json_schema_builder` code is unable to resolve the type annotations (like `datetime`) which is included in your sample model class (`Data`). This is happening because `get_type_hints` is called with the globals() and locals() of the `kernel_json_schema_builder` module, which do not include the types defined or imported in an outside module. As a result, it cannot find `datetime` and raises a NameError.

I modified the `kernel_json_schema_builder` class to pass the global namespace of the module where the model is defined `(vars(sys.modules[model.__module__]))` to the `get_type_hints` function. This makes it so that all the types used in the model's type annotations are available for resolution, thus fixing the error.

I will include this fix with the other type hint fix in a PR shortly.

**Comment by moonbox3** [comment]
Note, with the latest fix, when I run your sample code (thanks so much for including), I see:

```text
INFO:semantic_kernel.processes.local_runtime.local_step:Received message from `c5c77eb2fe2d49a894bc226e48fd20a8` targeting function `process_event` and parameters `data: timestamp=datetime.datetime(2024, 11, 7, 10, 22, 21, 210929)`.
INFO:semantic_kernel.processes.local_runtime.local_step:Step with Id '3cd4c42c7da44f69810e297c907b97a3' received all required input for function [process_event] and is executing.
INFO:semantic_kernel.processes.local_runtime.local_step:Invoking plugin `Step` and function `process_event` with arguments: {'data': Data(timestamp=datetime.datetime(2024, 11, 7, 10, 22, 21, 210929))}
INFO:semantic_kernel.functions.kernel_function:Function Step-process_event invoking.
INFO:root:Handled event with data timestamp=datetime.datetime(2024, 11, 7, 10, 22, 21, 210929)
INFO:semantic_kernel.functions.kernel_function:Function Step-process_event succeeded.
INFO:semantic_kernel.functions.kernel_function:Function completed. Duration: 0.000632s
```

**Comment by Druid-of-Luhn** [comment]
Wow, thanks for the quick resolution.

**Comment by moonbox3** [comment]
This fix is included in the recently released v1.14.0 package.
